crates/stark-felt/src/lib.rs

```diff
@@ -130,73 +161,223 @@ impl Felt {
+    /// assert_eq!(q, Felt::from(3));
+    /// assert_eq!(r, Felt::from(1));
+    /// ```
     pub fn div_rem(&self, rhs: &NonZeroFelt) -> (Self, Self) {
         let (q, r) = self.0.representative().div_rem(&rhs.0.representative());
         (Self(FieldElement::from(&q)), Self(FieldElement::from(&r)))
     }
 
-    /// Multiplicative inverse inside field.
+    /// Computes the multiplicative inverse of this [`Felt`] inside the field used by the Cairo VM.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use stark_felt::Felt;
+    /// let a = Felt::from(5);
+    /// let b = a.inverse();
+    /// assert_eq!(b.unwrap() * a, Felt::ONE);
+    /// ```
     pub fn inverse(&self) -> Option<Self> {
         self.0.inv().map(Self).ok()
     }
 
-    /// Finds the square root. There may be 2 roots for each square, and the lower one is returned.
-    pub fn sqrt(&self) -> Option<Self> {
+    /// Returns the square roots of this [`Felt`] within the field used by the Cairo VM.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use stark_felt::Felt;
+    /// let a = Felt::from(25);
+    /// let (b, c) = a.sqrts().unwrap();
+    /// assert_eq!(b * b, a);
+    /// assert_eq!(c * c, a);
+    ///
+    /// assert_eq!(Felt::from(24).sqrts(), None);
+    /// ```
+    pub fn sqrts(&self) -> Option<(Self, Self)> {
         let (root_1, root_2) = self.0.sqrt()?;
-        Some(Self(core::cmp::min(root_1, root_2)))
-    }
-
-    /// Raises `self` to the power of 2.
+        Some((Self(root_1), Self(root_2)))
+    }
+
+    /// Computes the square root of this [`Felt`]. There may be 2 roots for each square, and the
+    /// lower one is always returned.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use stark_felt::Felt;
+    /// let a = Felt::from(25);
+    /// let b = a.sqrt().unwrap();
+    ///
+    /// assert_eq!(b * b, a);
+    /// ```
+    pub fn sqrt(&self) -> Option<Self> {
+        self.sqrts().map(|(a, b)| a.min(b))
+    }
+
+    /// Raises this [`Felt`] to the power of 2.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use stark_felt::Felt;
+    /// let a = Felt::from(3);
+    /// let b = a.square();
+    /// assert_eq!(b, Felt::from(9));
+    /// ```   
     pub fn square(&self) -> Self {
         Self(self.0.square())
     }
 
-    /// Raises `self` to the power of `exponent`.
-    pub fn pow(&self, exponent: impl Into<u128>) -> Self {
-        Self(self.0.pow(exponent.into()))
-    }
-
-    /// Raises `self` to the power of `exponent`.
+    /// Raises this [`Felt`] to the power of `exponent`.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use stark_felt::Felt;
+    /// let a = Felt::from(3);
+    ///
+    /// assert_eq!(a.pow(2u32), Felt::from(9));
+    /// assert_eq!(a.pow(3u32), Felt::from(27));
+    /// assert_eq!(a.pow(0u32), Felt::ONE);
+    /// ```
+    pub fn pow(&self, exponent: impl IsUnsignedInteger) -> Self {
```

Comment:
This one would be a breaking change, and would also leak the lambdaworks dependency onto users.

Reply:
Oooh makes sense. I'll revert it 'cause it obviously falls out of the scope of the PR.

I feared converting to `u128` might trash registers if the function is not properly optimized by rustc.